crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -367,136 +404,184 @@ pub(super) struct FlatAbi {
     pub(super) align: u32,
 }
 
-/// Represents a pending event on a host-owned write end of a stream or future.
-///
-/// See `ComponentInstance::start_write_event_loop` for details.
-enum WriteEvent<B> {
-    /// Write the items in the specified buffer to the stream or future, and
-    /// return the result via the specified `Sender`.
-    Write {
-        buffer: B,
-        tx: oneshot::Sender<HostResult<B>>,
-    },
-    /// Drop the write end of the stream or future.
-    Drop(Option<Box<dyn FnOnce() -> B + Send + Sync>>),
-    /// Watch the read (i.e. opposite) end of this stream or future, dropping
-    /// the specified sender when it is dropped.
-    Watch { tx: oneshot::Sender<()> },
+/// Trait representing objects (such as streams, futures, or structs containing
+/// them) which require access to the store in order to be disposed of properly.
+pub trait DropWithStore: Sized {
+    /// Dispose of `self` using the specified store.
+    fn drop(self, store: impl AsContextMut) -> Result<()>;
+
+    /// Dispose of `self` using the specified accessor.
+    fn drop_with(self, accessor: impl AsAccessor) -> Result<()> {
+        accessor.as_accessor().with(move |store| self.drop(store))
+    }
 }
```

Comment:
WDYT about changing this to `&mut self`? I feel like that's a bit more idiomatic and maps well to `Drop`?

Also I think this trait will want to be clearly documented in its relation to `Drop`. Specifically this trait implementation does nothing outside of `WithAccessor` and that's required to do anything here.